### PR TITLE
SDK-269: Added ValidateConfigurationsMojo.

### PR DESCRIPTION
**Pull request by mks-d**
Ticket: https://issues.openmrs.org/browse/SDK-269

### CODE REVIEW

File: integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java
Comment (mks-d) [inline integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java:40]: @mseaton this the main reason this PR is still WIP is because I couldn't get this to work...
The instruction
```java
configProject.executeGoal("validate", "-N", "-X");
```
doesn't actually get to `ValidateConfigurationsMojo#execute`, any idea why? I'm not too sure how this Maven `Verifier` works.

File: pom.xml
Comment (mks-d) [inline pom.xml:157]: @mseaton is that ok? I needed this for the `Validator` actually, that is compiled against Java 8.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java
Comment (mks-d) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java:85]: @mseaton the biggest puzzle. Failing to do this triggers errors when the Spring context is destroyed. Everything works fine otherwise but the logs get bloated with weird linkage errors (if I remember correctly.)

**Review by mks-d — COMMENTED** [review]
@mseaton this is working fine but I left a couple of comments.

File: pom.xml
Comment (mseaton) [reply pom.xml:157]: @mks-d yes, this should be no problem.  I'll try to look over / investigate the rest of the PR soon.

File: integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java
Comment (mseaton) [reply integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java:40]: @mks-d , it's possible I'm not doing things exactly right, but what I've found is that you need to add the executions you want to run explicitly into the pom.  So, here you are trying to use the "config-test-parent" pom to test the project.  That pom is found in src/test/resources/config-test-parent/pom.xml.  You'll need to add an execution to the end of that, something like this:

```                    
<execution>
  <id>validate-configurations</id>
  <phase>validate</phase>
    <goals>
    <goal>validate-configurations</goal>
  </goals>
</execution>
```

File: integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java
Comment (mks-d) [inline integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java:28]: @mseaton I completely turned this test around. As a matter of fact it is not anymore a Mojo _integration_ test, so happy to move it in the regular test files location. Just let me know your thoughts.

**Review by mks-d — COMMENTED** [review]
@mseaton of course this is not ready to go until Iniz' Validator JAR is properly Maven deployed, but otherwise it's working as far as I can tell.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java
Comment (mseaton) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java:87]: @mks-d I can try to poke around, but nothing springs immediately to mind.  Would be nice to figure out and clean up but shouldn't need to hold up the merge and getting this into the project.  As soon as the validator jar is generally available in iniz, and this will build cleanly, it looks good to me.

I wasn't able to test this breaking.  If I comment out the findClassDefinitions(); line and run a mvn clean install, everything builds and all unit tests pass.  Is there a way to reproduce?

Comment (mks-d) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java:87]: Thanks @mseaton, much appreciated.

>Is there a way to reproduce?

This is a runtime-only issue, not a build issue. If you configure the validate goal for a config in its POM and that you comment out `findClassDefinitions()` you should see the problem coming up.

Comment (mseaton) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java:87]: OK, will see if I can try it at some point.  I'll leave it to you to merge this in when ready based on when the validator jar is ready and available for this to build with.

**Comment by Ruhanga** [comment]
Hi @ibacher, @mseaton, could take a look when you've gotten the chance and confirm whether we are good to have this merged in? Thanks.

**Review by mseaton — APPROVED** [review]
Assuming this has been tested and works and meets your needs, I'm supportive of merging this in.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java
Comment (Ruhanga) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java:45]: I'd leave `compile-configurations` as is to harmonise with the existing `goal`s' naming convention.

Comment (Ruhanga) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsMojo.java:45]: I'd leave `compile-configurations` as is to harmonise with the existing `goal`s' naming convention.

**Review by Ruhanga — COMMENTED** [review]
@ibacher, thanks for catch on `java.lang.NoClassDefFoundError`.  This is most likely resolved by adding a corresponding line (of the supposedly missing definition) to the list discussed here: https://github.com/openmrs/openmrs-contrib-packager-maven-plugin/pull/7#discussion_r575299459

File: integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java
Comment (Ruhanga) [reply integration-tests/src/test/java/org/openmrs/maven/plugins/packager/config/ValidateConfigurationsTest.java:28]: @ibacher, since this is taking advantage of the already existing test resources, we can leave it here I guess?

**Comment by ibacher** [comment]
Thanks for the research @Ruhanga! Plexus really is a PIA.

Do we have some kind of solution for the logging thing I mentioned in the review above? To me, that's much more critical than the rest of it to getting this mergable.

**Comment by Ruhanga** [comment]
>Do we have some kind of solution for the logging thing I mentioned in the review above? 

Oh, sorry I lost focus on the above. This should be possible using the `initializer.logging.location` ([doc](https://github.com/mekomsolutions/openmrs-module-initializer/blob/master/readme/rtprops.md#2-initializerlogginglocation-optional)) runtime prop. I'll play around with it and add what works.

**Comment by Ruhanga** [comment]
>Do we have some kind of solution for the logging thing I mentioned in the review above?

@ibacher, I've just noticed this already supported with the following execution configuration which is fair enough I guess;
```
<configuration>
    <extraValidatorArgs>--log-dir=${project.build.directory}</extraValidatorArgs>
</configuration>
```

**Comment by ibacher** [comment]
Sure; can that be made the default behaviour, though? It’s not great that the log (which is really the important output of the validator) isn’t immediately obvious.

**Comment by Ruhanga** [comment]
Done in the the latest commit @ibacher. Could you review again?

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
LGTM. Thanks @Ruhanga!

**Comment by Ruhanga** [comment]
Hi @ibacher, @mks-d, if all is good, can we merge this in? Thanks.

**Pull request merged**
